File: 1575-count-all-possible-routes/1575-count-all-possible-routes.py
class Solution:
    def solve(self, curr_city: int, remain_fuel: int, finish: int, memo: Dict[Tuple[int], int], locations: List[int]) -> int:
        if remain_fuel < 0:
            return 0
        if (curr_city, remain_fuel) in memo:
            return memo[(curr_city, remain_fuel)]
        
        ans = 0
        if curr_city == finish:
            ans = 1
        
        for next_city in range(self.n):
            if next_city != curr_city:
                ans = (ans + self.solve(next_city, remain_fuel - abs(locations[curr_city] - locations[next_city]), finish, memo, locations)) % 1000000007
        
        memo[(curr_city, remain_fuel)] = ans
        return ans
    
    def countRoutes(self, locations: List[int], start: int, finish: int, fuel: int) -> int:
        ans, self.n = 0, len(locations)
        
        memo = dict()
        return self.solve(start, fuel, finish, memo, locations)
        
        

File: 1575-count-all-possible-routes/1575-count-all-possible-routes.cpp
class Solution {
private:
    int n;
    int solve(vector<int>& locations, int currCity, int finish, int remainingFuel, vector<vector<int>>& memo){
        if (remainingFuel < 0) return 0;
        if (memo[currCity][remainingFuel] != -1) return memo[currCity][remainingFuel];
        
        int ans = currCity == finish? 1: 0;
        for (int nextCity = 0; nextCity < n; ++nextCity){
            if (nextCity != currCity){
                ans = (ans + solve(locations, nextCity, finish, remainingFuel - abs(locations[currCity] - locations[nextCity]), memo)) % 1'000'000'007;
            }
        }
        
        return memo[currCity][remainingFuel] = ans;
    }
    
public:
    int countRoutes(vector<int>& locations, int start, int finish, int fuel) {
        this -> n = locations.size();
        vector<vector<int>> memo(n, vector<int>(fuel + 1, -1));
        
        return solve(locations, start, finish, fuel, memo);
    }
};